vmnics.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"fmt"
+	"math/rand"
+	"strings"
+
+	"github.com/apex/log"
+	"github.com/pkg/errors"
+)
+
+const randChars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+
+func randomString(length int) string {
+	if length == 0 {
+		length = 8
+	}
+	randCharLength := len(randChars)
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = randChars[rand.Intn(randCharLength)]
+	}
+	return string(b)
+}
+
+const LinuxIfnameLength int = 15
+
+// generate a random Linux network interface name.
+// Allow specifying a prefix
+func randomNicIFName(prefix string) string {
+	prefixLength := len(prefix)
+	if prefixLength >= LinuxIfnameLength {
+		return randomString(LinuxIfnameLength)
+	}
+	randLength := LinuxIfnameLength - prefixLength
+	return prefix + randomString(randLength)
+}
+
+// The VMNic lifecycle: first you create one with vm.AddNic().
+// The you call Setup on it to create the tap.  Then you call
+// Args to get the qemu commandline to use it.  Finally you
+// call Cleanup to tear down the nic.
+type VMNic struct {
+	BusAddr    string
+	DeviceType string
+	HWAddr     string
+	ID         string
+	IFName     string
+	NetIFName  string
+	NetType    string
+	NetAddr    string
+	BootIndex  string
+	Ports      []PortRule
+}
+
+func (v *VM) AddNic(nicdef *NicDef, netdef *NetworkDef) error {
+	log.Debugf("AddNic: Nic: %v on Network %v", nicdef, netdef)
+	nicdef.IFName = randomNicIFName("anic")
+	nic := VMNic{
+		BusAddr:    nicdef.BusAddr,
+		DeviceType: nicdef.Device,
+		HWAddr:     nicdef.Mac,
+		ID:         nicdef.ID,
+		IFName:     nicdef.IFName,
+		NetIFName:  netdef.IFName,
+		NetType:    netdef.Type,
+		NetAddr:    netdef.Address,
+		BootIndex:  nicdef.BootIndex,
+		Ports:      nicdef.Ports,
+	}
+	log.Debugf("AddNic: Nic:%s Slot:%s -> IFName:%s Type:%s", nic.ID, nic.BusAddr, nic.IFName, nic.NetType)
+	v.opts.NICs = append(v.opts.NICs, &nic)
+	return nil
+}
+
+var macOffset int = 0
```

Comment:
Just a note that this works well so long as we continue to start clusters as a whole, but won't work if we start having separate programs start the nics.  (Unless we use an API to request a single daemon set up the hwaddrs)

Reply:
Yes, I think we should:

1) switch to starting clusters-as-a-whole via API
2) enable API to start/stop VMs within a cluster

Reply:
Sounds good.